rrdd/rrdd_server.ml

```diff
@@ -275,18 +275,69 @@ let forget_host_ds _ ~(ds_name : string) : unit =
 	)
 
 let query_possible_dss rrdi =
-	let enabled_dss = Rrd.ds_names rrdi.rrd in
-	let open Ds in
-	let open Data_source in
-	List.map (fun ds -> {
-		name = ds.ds_name;
-		description = ds.ds_description;
-		enabled = List.mem ds.ds_name enabled_dss;
-		standard = ds.ds_default;
-		min = ds.ds_min;
-		max = ds.ds_max;
-		units = ds.ds_units;
-	}) rrdi.dss
+  (* We have data sources coming from different places, so we want the union of these;
+     this happens to be the union of the rrdi.rrd (live) and rrdi.rrd.rdd_dss (archival).
+     This is straighforward, but those two sets of ds have different types,
+     hence we need to coerce them into a common type (this is why we have
+     'description not available', and 'units unknown' etc.).
+     Deciding whether a ds is enabled is also not obvious. If we have a 'live'
+     ds, then it is enabled if it exists in the set rrdi.rrd. If we have an 'archival'
+     ds, then it is enabled if it is also an enabled 'live' ds, otherwise it is
+     disabled.
+   *)
+  let module SSet = Set.Make(String) in
+  let module SMap = Map.Make(String) in
+  let open Ds in
+  let open Data_source in
+
+  let live_sources = rrdi.dss in
+  let archival_sources = rrdi.rrd.rrd_dss in
+
+  let name_to_live_dss =
+    let enabled_names = Rrd.ds_names rrdi.rrd |> SSet.of_list in
+    let is_live_ds_enabled ds = SSet.mem ds.ds_name enabled_names in
+    live_sources |>
+    List.fold_left (fun map ds ->
+      let k = ds.ds_name in
+      let v = { name = ds.ds_name
+              ; description = ds.ds_description
+              ; enabled = is_live_ds_enabled ds
+              ; standard = ds.ds_default
+              ; min= ds.ds_min
+              ; max = ds.ds_max
+              ; units = ds.ds_units
+              }
+      in
+      SMap.add k v map) SMap.empty
+  in
+  let name_to_disabled_dss =
+    archival_sources |>
+    Array.fold_left (fun map ds ->
+      if SMap.mem ds.Rrd.ds_name name_to_live_dss then
+        map
+      else
+        let k = ds.ds_name in
+        let v = { name = ds.ds_name
+                ; description = "description not available"
+                ; enabled = false
+                ; standard = false
+                ; units = "unknown"
+                ; min = ds.ds_min
+                ; max = ds.ds_max
+                }
+        in
+        SMap.add k v map) SMap.empty
+  in
+  let merge_f k mv1 mv2 = match mv1, mv2 with
+  | None, None       -> None
+  | Some v1, None    -> Some v1
+  | None, Some v2    -> D.error "DEBUG disabled key, k: %s" k; Some v2
+  | Some v1, Some v2 -> failwith "impossible: live and archival disabled keys should be disjoint"
```

Comment:
We probably want to log an error instead of failing horribly. I know this case is checked against in lines 316-317 but you never know what will happen to the code long term...

Reply:
i disagree, since I don't trust we would notice the bug if somebody were to make it so that live and archival are no longer disjoint and an error was not raised

Reply:
to clarify, i _do_ think we would notice with `failwith`

Reply:
My worry is that raising an exception will stop the http server altogether until the daemon is restarted. My point is that it's preferable to let a petition fail and log it rather than taking the service down.

Thinking about this further I think we should avoid a merge than can raise conflicts, the original patch used this:

```
    let add_seq i m =
      Seq.fold_left (fun m (k,v) -> add k v m) m i
```
Simply adding the key values from the second map to the first will match the behaviour in master.

Reply:
i can get on board with this